src/random_bq_to_anndata.py

```diff
@@ -1,46 +1,178 @@
+    query = client.query(f"SELECT MIN(cas_cell_index) AS min_cas_cell_index, MAX(cas_cell_index) AS max_cas_cell_index FROM `{project}.{dataset}.cas_cell_info`")
+    row = list(query.result())[0]
+    min_cas_cell_index, max_cas_cell_index = row.min_cas_cell_index, row.max_cas_cell_index
+    print(f"Getting {num_cells} random IDs between {min_cas_cell_index} and {max_cas_cell_index}...")
+    cell_ids = list(range(min_cas_cell_index, max_cas_cell_index + 1))
     random.shuffle(cell_ids)
     del cell_ids[num_cells:]
+    print(f"Random IDs: {cell_ids}")
     return cell_ids
 
+# Retrieve all of the features from the database
+def get_features(project, dataset, client):
+    sql = f"SELECT cas_feature_index, original_feature_id, feature_name FROM `{project}.{dataset}.cas_feature_info` ORDER BY cas_feature_index"
+    query = client.query(sql)
+    ordered_feature_indices = []
+    feature_index_to_feature = {}
+    for row in query:
+        index = row["cas_feature_index"]
+        ordered_feature_indices.append(index)
+        feature_index_to_feature[index] = Feature(row["cas_feature_index"], row["original_feature_id"], row["feature_name"])
+    return ordered_feature_indices, feature_index_to_feature
 
-def get_cell_data(project, dataset, client, num_cells):
-    random_ids = get_random_ids(project, dataset, client, num_cells)
-    in_clause = f" matrix.cas_cell_index IN ({','.join(map(str, random_ids))})"
+# Retrieve metadata for the selected (random) cells from the database
+def get_cells(project, dataset, client, random_cell_ids):
+    in_clause = f" cas_cell_index IN ({','.join(map(str, random_cell_ids))})"
+    sql = f"SELECT cas_cell_index, original_cell_id, cell_type FROM `{project}.{dataset}.cas_cell_info` WHERE " + in_clause + " ORDER BY cas_cell_index"
+    query = client.query(sql)
+    ordered_cell_indices = []
+    cell_index_to_cell = {}
+    for row in query:
+        index = row["cas_cell_index"]
+        ordered_cell_indices.append(index)
+        cell_index_to_cell[index] = Cell(row["cas_cell_index"], row["original_cell_id"], row["cell_type"])
+    return ordered_cell_indices, cell_index_to_cell
+
+def get_matrix_data(project, dataset, client, random_cell_ids):
+    in_clause = f" matrix.cas_cell_index IN ({','.join(map(str, random_cell_ids))})"
 
     # at some point, we will probably want create temp table of cell_ids and then JOIN on it
     # instead of an IN clause
-    sql = f"SELECT original_cell_id, cell_type, original_feature_id, feature_name, raw_counts AS count FROM `{project}.{dataset}.cas_cell_info` AS cell, `{project}.{dataset}.cas_feature_info` AS feature, `{project}.{dataset}.cas_raw_count_matrix` AS matrix WHERE matrix.cas_cell_index = cell.cas_cell_index AND matrix.cas_feature_index = feature.cas_feature_index AND" + in_clause
-    print(f"Getting {num_cells} random cells' data from {project}.{dataset}...")
+    # NOTE - really don't need to join to cas_cell_index and cas_feature_index here anymore.
```

Comment:
any reason not to remove these joins?

Reply:
I've removed the join on features. I'm trying to convince myself that it's safe to remove that to cell. But we *could* have a case where one of the cas_cell_ids is NOT in the matrix.

Probably going to rewrite it to have the random ids be pulled from the matrix itself.

Reply:
Pulling from the matrix is going to be extremely expensive… It's valid (though strange) to have a cell_id that has no row in the matrix.  It means that all the values were zero (since this is a sparse matrix).  We should never have a value in the matrix thought that does't have a corresponding entry in the cell table.

Reply:
I've removed the joins. The code handles this case (where a cell that is present in cas_cell_index has no data in the matrix).